### ISSUE TITLE
Slow Indigo Woodpecker - Oracle.sol: observe function has overflow risk and should cast to uint256 like Uniswap V3 does

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-10-aloe-judging#85]
Slow Indigo Woodpecker

medium

# Oracle.sol: observe function has overflow risk and should cast to uint256 like Uniswap V3 does
## Summary
The `Oracle.observe` function basically uses the same math from the Uniswap V3 code to search for observations.  

In comparison to Uniswap V3, the `Oracle.observe` function takes a `seed` such that the runtime of the function can be decreased by calculating the `seed` off-chain to act as a hint for finding the observation.  

In the process of copying the Uniswap V3 code, a `uint256` cast has been forgotten which introduces a risk of intermediate overflow in the `Oracle.observe` function.  

Thereby the `secondsPerLiquidityCumulativeX128` return value can be wrong which can corrupt the implied volatility (ÌV) calculation.  

## Vulnerability Detail
Looking at the `Oracle.observe` function, the `secondsPerLiquidityCumulativeX128` return value is calculated as follows:

https://github.com/aloelabs/aloe-ii/blob/c71e7b0cfdec830b1f054486dfe9d58ce407c7a4/core/src/libraries/Oracle.sol#L196
```solidity
liqCumL + uint160(((liqCumR - liqCumL) * delta) / denom)
```

The calculation is done in an `unchecked` block. `liqCumR` and `liqCumL` have type `uint160`.  
`delta` and `denom` have type `uint56`.  

Let's compare this to the Uniswap V3 code.

https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/libraries/Oracle.sol#L279-L284
```solidity
beforeOrAt.secondsPerLiquidityCumulativeX128 +
    uint160(
        (uint256(
            atOrAfter.secondsPerLiquidityCumulativeX128 - beforeOrAt.secondsPerLiquidityCumulativeX128
        ) * targetDelta) / observationTimeDelta
    )
```

The result of `atOrAfter.secondsPerLiquidityCumulativeX128 - beforeOrAt.secondsPerLiquidityCumulativeX128` is cast to `uint256`.  

That's because multiplying the result by `targetDelta` can overflow the `uint160` type.  

The maximum value of `uint160` is roughly `1.5e48`.  

`delta` is simply the time difference between `timeL` and `target` in seconds.  

The `secondsPerLiquidityCumulative` values are accumulators that are calculated as follows:
https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/libraries/Oracle.sol#L41-L42
```solidity
secondsPerLiquidityCumulativeX128: last.secondsPerLiquidityCumulativeX128 +
    ((uint160(delta) << 128) / (liquidity > 0 ? liquidity : 1)),
```

If `liquidity` is very low and the time difference between observations is very big (hours to days), this can lead to the intermediate overflow in the `Oracle` library, such that the `secondsPerLiquidityCumulative` is much smaller than it should be.  

The lowest value for the above division is `1`. In that case the accumulator grows by `2^128` (`~3.4e38`) every second.

If observations are apart 24 hours (`86400 seconds`), this can lead to an overflow:
Assume for simplicity `target - timeL = timeR - timeL`
```text
(liqCumR - liqCumL) * delta = 3.4e38 * 86400 * 86400 > 1.5e48`
```

## Impact
The corrupted return value affects the [`Volatility` library](https://github.com/aloelabs/aloe-ii/blob/c71e7b0cfdec830b1f054486dfe9d58ce407c7a4/core/src/libraries/Volatility.sol#L121). Specifically, the IV calculation.    

This can lead to wrong IV updates and LTV ratios that do not reflect the true IV, making the application more prone to bad debt or reducing capital efficiency.  

## Code Snippet
https://github.com/aloelabs/aloe-ii/blob/c71e7b0cfdec830b1f054486dfe9d58ce407c7a4/core/src/libraries/Oracle.sol#L196

https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/libraries/Oracle.sol#L279-L284

## Tool used
Manual Review

## Recommendation
Perform the same cast to `uint256` that Uniswap V3 performs:  
```solidity
liqCumL + uint160((uint256(liqCumR - liqCumL) * delta) / denom)
```

**Comment by panprog** [comment]
Escalate

This should be a valid medium, not high, because while the situation is possible, it can only happen when liquidity is extremely low (basically no liquidity). The example calculation assumes minimum liquidity value (0 or 1):
```solidity
secondsPerLiquidityCumulativeX128: last.secondsPerLiquidityCumulativeX128 +
    ((uint160(delta) << 128) / (liquidity > 0 ? liquidity : 1)),
...
(liqCumR - liqCumL) * delta = 3.4e38 * 86400 * 86400 ~= 2.5e48 > 1.5e48`
```

Even with such a low liquidity value the overflow happens just barely. Even a value of 2 for liquidity already won't overflow. If the pool has 0 liquidity (or 1, 2 or even 100 - even for token with 6 decimals that's basically empty pool) for over a day - nobody should/will really use it. Any semi-active pool which will actually be used will have much higher liquidity at all times.

So the issue describes a very edge case which most probably won't ever happen in real life.

The reporter probably also understands it and set his reported issue severity to medium, not high.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This should be a valid medium, not high, because while the situation is possible, it can only happen when liquidity is extremely low (basically no liquidity). The example calculation assumes minimum liquidity value (0 or 1):
> ```solidity
> secondsPerLiquidityCumulativeX128: last.secondsPerLiquidityCumulativeX128 +
>     ((uint160(delta) << 128) / (liquidity > 0 ? liquidity : 1)),
> ...
> (liqCumR - liqCumL) * delta = 3.4e38 * 86400 * 86400 ~= 2.5e48 > 1.5e48`
> ```
> 
> Even with such a low liquidity value the overflow happens just barely. Even a value of 2 for liquidity already won't overflow. If the pool has 0 liquidity (or 1, 2 or even 100 - even for token with 6 decimals that's basically empty pool) for over a day - nobody should/will really use it. Any semi-active pool which will actually be used will have much higher liquidity at all times.
> 
> So the issue describes a very edge case which most probably won't ever happen in real life.
> 
> The reporter probably also understands it and set his reported issue severity to medium, not high.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by roguereddwarf** [comment]
@panprog 

I agree with the escalation. I intentionally reported this as Medium due to the Low likelihood.

**Comment by Trumpero** [comment]
Planning to accept escalation and downgrade this issue to medium because of the low likelihood.

**Comment by Czar102** [comment]
Result:
Medium
Unique

All parties agree on the above result.

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [panprog](https://github.com/sherlock-audit/2023-10-aloe-judging/issues/85/#issuecomment-1801857655): accepted

**Comment by haydenshively** [comment]
Fixed in https://github.com/aloelabs/aloe-ii/pull/208

**Comment by roguereddwarf** [comment]
Mitigation Review:

The issue has been fixed by applying the recommended `uint256` cast that is also used by Uniswap V3's own Oracle implementation.
Now there won't be an intermediate overflow even for very low liquidity pools.
